src/shared/test/port-provider-test.js

```diff
@@ -0,0 +1,285 @@
+import { delay } from '../../test-util/wait';
+import { PortProvider } from '../port-provider';
+
+const source = 'hypothesis';
+
+describe('PortProvider', () => {
+  let portProvider;
+
+  function portFinderRequest({
+    data,
+    origin = window.location.origin,
+    source = window,
+  }) {
+    const event = new MessageEvent('message', {
+      data,
+      origin,
+      source,
+    });
+
+    window.dispatchEvent(event);
+
+    return event;
+  }
+
+  beforeEach(() => {
+    sinon.stub(window, 'postMessage');
+    portProvider = new PortProvider(window.location.origin);
+    portProvider.listen();
+  });
+
+  afterEach(() => {
+    window.postMessage.restore();
+    portProvider.destroy();
+  });
+
+  describe('#destroy', () => {
+    it('ignores valid port request if `PortFinder` has been destroyed', async () => {
+      portProvider.destroy();
+      portFinderRequest({
+        data: {
+          channel: 'host-sidebar',
+          port: 'sidebar',
+          source,
+          type: 'request',
+        },
+      });
+      await delay(0);
+
+      assert.notCalled(window.postMessage);
+    });
+  });
+
+  describe('#getPort', () => {
+    it('returns `null` if called with wrong arguments', () => {
+      let hostPort;
+
+      // Incorrect channel
+      hostPort = portProvider.getPort({
+        channel: 'notebook-sidebar',
+        port: 'host',
+      });
+      assert.isNull(hostPort);
+
+      // Incorrect port
+      hostPort = portProvider.getPort({
+        channel: 'host-sidebar',
+        port: 'sidebar',
+      });
+      assert.isNull(hostPort);
+    });
+
+    it('returns the `host` port of the `host-sidebar` channel if called with the right arguments', () => {
+      const hostPort = portProvider.getPort({
+        channel: 'host-sidebar',
+        port: 'host',
+      });
+      assert.exists(hostPort);
+    });
+  });
+
+  describe('#listen', () => {
+    it('ignores all port requests before `listen` is called', async () => {
```

Comment:
I was looking for a `listen` call in this test, but then realized it was in the setup code. I think it might make sense to call `listen` explicitly in each test, even though it will lead to more lines of code in total.